### PR TITLE
Fixed slow rest api to get boot parameters

### PR DESCRIPTION
**Pull request by shunr-hpe**
Made the following improvements when getting boot parameters by name,
nid, or mac.
- Changed to first look up the name directly in the etcd data. In
  most cases the request is going to be for an x name and so this
  check will be the fastest way to satisfy the request.
- Changed to use the values it gets in its initial getting of etcd
  data. The previous implementation was only using the names,
  and then re-getting the value on a per name case.
- Changed to only get the kernel and initrd data once per http
  request.
- Added an "ID to SMComponent" map for the cached state
  manager data. The previous implementation was looping
  through this data for each boot parameter entry.

Jira: CASMHMS-4540

### Summary and Scope

HTTP calls like following are made often in some install scripts.
/boot/v1/bootparameters?name=x3001c0s31b0n0

This call is slow when there are a lot of boot parameters. The reason it is slow is because bss retrieves the etcd data multiple times and loops through the state manager data multiple times. 

### Issues and Related PRs

* Resolves CASMHMS-4540

### Testing

Tested on wasp. Ran the smoke and integration tests. Compared the results of the following calls before and after the changes.
```
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters'
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/dumpstate'
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters?name=x3000c0s10b2n0'
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters?nid=100004'
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters?mac=98:03:9b:aa:97:20'
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters?name=x3000c0s10b2n0,x3000c0s1b0n0&mac=48:df:37:b2:3c:0f,48:df:37:b2:3c:1f,junk,98:03:9b:aa:97:20&nid=100004,42'

# in smd but not boot parameters
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters?name=x3000c0s10b4n0'

# does not exist in bss and smd
curl -s -X GET -H "Authorization: Bearer ${TOKEN}" 'https://api-gw-service-nmn.local/apis/bss/boot/v1/bootparameters?name=x3000c0s10b4n1'
```

Tested in docker compose environment. Verified that the results are the same before and after these changes.
http get:

- /boot/v1/bootparameters?nid=100020
- /boot/v1/bootparameters?mac=ff:ff:ff:ff:ff:ff
- /boot/v1/bootparameters?mac=b8:59:9f:d9:9d:68,b8:59:9f:d9:9d:69,b8:59:9f:d9:9d:b1
- /boot/v1/bootparameters?name=x4499c0s42b0n0,x3000c0s36b0n0,x3000c0s36b0n0junk,x3000c0s36b0&mac=a4:bf:01:28:9b:17,junk&nid=1,49169408,999999
- /boot/v1/bootparameters?name=x3001c0s31b0n0
- DELETE /boot/v1/bootparameters -d '{ "hosts" : [ "x3001c0s31b0n0" ] }; GET /boot/v1/bootparameters?name=x3001c0s31b0n0; GET /boot/v1/bootparameters?nid=100016'
- Tested fqdn names by changing the code to inject them into the cached smd data.

Were the install/upgrade based validation checks/tests run?(goss tests/install-validation doc)
Were continuous integration tests run? Y
Was an Upgrade tested?                 Y
Was a Downgrade tested?                Y
If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?

### Risks and Mitigations

A different method of searching is being used. There could be differences. To mitigate I tested multiple different cases.

### CODE REVIEW

File: cmd/boot-script-service/boot_data.go [hunk 1]
```diff
@@ -819,6 +870,26 @@ func LookupGlobalData() (BootData, error) {
 	return LookupByRole(GlobalTag)
 }
 
+func LookupComponentByName(name string) SMComponent {
+	comp, _ := FindSMCompByNameInCache(name)
+	return comp
+}
+
+func ToBootData(value string, kernalImages map[string]ImageData, initrdImages map[string]ImageData) (BootData, error) {
```
Comment (rsjostrand-hpe) [inline cmd/boot-script-service/boot_data.go:878]: Small typo `kernal` ->`kernel`

File: cmd/boot-script-service/boot_data.go [hunk 2]
```diff
@@ -787,6 +810,34 @@ func lookup(name, altName, role, defaultTag string) BootData {
 	return bd
 }
 
+func bdConvertUsingImageCache(bds BootDataStore, kernelImages map[string]ImageData, initrdImages map[string]ImageData) (ret BootData) {
```
Comment (rsjostrand-hpe) [inline cmd/boot-script-service/boot_data.go:813]: Small typo `kernal` -> `kernel`

File: cmd/boot-script-service/default_api.go [hunk 3]
```diff
@@ -239,44 +240,72 @@ func BootparametersGet(w http.ResponseWriter, r *http.Request) {
...
-					}
-				}
+	}
+	args.Hosts = unfoundHosts
+
+	if len(args.Hosts) > 0 || len(args.Macs) > 0 || len(args.Nids) > 0 {
+
+		nameValues := GetNamesAndValues()
+
+		kernalImages := make(map[string]ImageData)
```
Comment (rsjostrand-hpe) [inline cmd/boot-script-service/default_api.go:264]: Small typo `kernal`-> `kernel`

File: cmd/boot-script-service/boot_data.go (hunk 1 above)
Comment (shunr-hpe) [reply cmd/boot-script-service/boot_data.go]: Thanks

**Review by nieuwsma — APPROVED** [review]

**Review by mbuchmann-hpe — APPROVED** [review]

**Pull request merged**